components/sup-client/src/lib.rs

```diff
@@ -140,56 +126,48 @@ impl From<termcolor::ParseColorError> for SrvClientError {
-    ///
-    /// Disconnect by simply dropping all instances of the `SrvClient`.
+    /// Connect to the remote server with the given secret_key and make a request.
     ///
-    /// *Note:* The client does not contain any reconnection logic. If the socket becomes
-    /// disconnected all future requests will fail.
-    pub async fn connect(address: &ListenCtlAddr,
-                         secret_key: &str)
-                         -> Result<SrvClient, SrvClientError> {
-        let socket = TcpStream::connect(address.as_ref()).await?;
-        let (reader, writer) = tokio_io::split(socket);
-        let reader = FramedRead::new(reader, SrvCodec::new());
-        let writer = FramedWrite::new(writer, SrvCodec::new());
-        let (disconnect_tx, disconnect_rx) = oneshot::channel();
-        let sync = SyncSrvClient { writer,
-                                   current_transaction: SrvTxn::default(),
-                                   transaction_lookup: HashMap::new(),
-                                   // Default to connected. This instance will only be returned
-                                   // if the connect is successfull.
-                                   state: SrvClientState::Connected,
-                                   disconnect_tx: Some(disconnect_tx) };
-        let mut client = SrvClient { sync: Arc::new(Mutex::new(sync)), };
-
-        tokio::spawn(Self::handle_messages(Self::clone(&client), reader, disconnect_rx));
+    /// Returns a stream of `SrvMessage`'s representing the server response.
+    pub async fn request(
+        address: &ListenCtlAddr,
+        secret_key: &str,
+        request: impl Into<SrvMessage> + fmt::Debug)
+        -> Result<impl Stream<Item = Result<SrvMessage, io::Error>>, SrvClientError> {
+        println!("IN NEW request");
 
-        // Handshake with the server
-        let mut request = protocol::ctl::Handshake::default();
-        request.secret_key = Some(String::from(secret_key));
-        let mut response = client.request(request).await?;
-        // Verify there was only one response message and it was not an error
-        response.recv()
-                .await
-                .ok_or(SrvClientError::ConnectionClosed)??;
-        debug_assert!(response.recv().await.is_none());
+        let socket = TcpStream::connect(address.as_ref()).await?;
+        let mut socket = Framed::new(socket, SrvCodec::new());
+        let mut current_transaction = SrvTxn::default();
+
+        // Send the handshake message to the server
+        let mut handshake = protocol::ctl::Handshake::default();
+        handshake.secret_key = Some(String::from(secret_key));
+        let mut message = SrvMessage::from(handshake);
+        message.set_transaction(current_transaction);
+        socket.send(message).await?;
+
+        // Verify the handshake response
+        socket.next()
+              .await
+              .ok_or(SrvClientError::ConnectionClosed)??;
```

Comment:
Do we need some kind of timeout for this, or is that effectively handled on the server-side?

Reply:
The server should handle it, but it would probably be good to add. Addressed in [e9305e](https://github.com/habitat-sh/habitat/pull/7256/commits/e9305ebf5782036ede72505e862e4ea3525669f9).